internal/fanoutconsumer/traces.go

```diff
@@ -8,34 +8,38 @@ import (
 
 	"go.uber.org/multierr"
 
+	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/pdata/ptrace"
 )
 
 // NewTraces wraps multiple trace consumers in a single one.
 // It fanouts the incoming data to all the consumers, and does smart routing:
 //   - Clones only to the consumer that needs to mutate the data.
 //   - If all consumers needs to mutate the data one will get the original mutable data.
-func NewTraces(tcs []consumer.Traces) consumer.Traces {
+func NewTraces(tcs map[component.ID]consumer.Traces) consumer.Traces {
 	// Don't wrap if there is only one non-mutating consumer.
```

Comment:
I put this in the fanoutconsumer because it's frequently used in the graph, but ideally we would have some universal wrapper that handles this for most components.